qiskit_nature/problems/sampling/protein_folding/bead_distances/distance_map_builder.py

```diff
@@ -0,0 +1,317 @@
+                    )
+                ).reduce()
+                num_distances += 1
+            except KeyError:
+                pass
+            try:
+                distance_map[lower_side_bead][upper_main_bead] = _fix_qubits(
+                    (
+                        delta_n0[lower_bead_ind][1][upper_bead_ind][0] ** 2
+                        + delta_n1[lower_bead_ind][1][upper_bead_ind][0] ** 2
+                        + delta_n2[lower_bead_ind][1][upper_bead_ind][0] ** 2
+                        + delta_n3[lower_bead_ind][1][upper_bead_ind][0] ** 2
+                    )
+                ).reduce()
+                num_distances += 1
+            except KeyError:
+                pass
+            try:
+                distance_map[lower_main_bead][upper_side_bead] = _fix_qubits(
+                    (
+                        delta_n0[lower_bead_ind][0][upper_bead_ind][1] ** 2
+                        + delta_n1[lower_bead_ind][0][upper_bead_ind][1] ** 2
+                        + delta_n2[lower_bead_ind][0][upper_bead_ind][1] ** 2
+                        + delta_n3[lower_bead_ind][0][upper_bead_ind][1] ** 2
+                    )
+                ).reduce()
+                num_distances += 1
+            except KeyError:
+                pass
+            try:
+                distance_map[lower_side_bead][upper_side_bead] = _fix_qubits(
+                    (
+                        delta_n0[lower_bead_ind][1][upper_bead_ind][1] ** 2
+                        + delta_n1[lower_bead_ind][1][upper_bead_ind][1] ** 2
+                        + delta_n2[lower_bead_ind][1][upper_bead_ind][1] ** 2
+                        + delta_n3[lower_bead_ind][1][upper_bead_ind][1] ** 2
+                    )
+                ).reduce()
+                num_distances += 1
+            except KeyError:
+                pass
+
+    logger.info(num_distances, " distances created")
+    return distance_map, num_distances
+
+
+def _calc_distances_main_chain(
+    peptide: Peptide,
+) -> Tuple[
+    DefaultDict[Any, DefaultDict[Any, DefaultDict[Any, dict]]],
+    DefaultDict[Any, DefaultDict[Any, DefaultDict[Any, dict]]],
+    DefaultDict[Any, DefaultDict[Any, DefaultDict[Any, dict]]],
+    DefaultDict[Any, DefaultDict[Any, DefaultDict[Any, dict]]],
+]:
+    """
+    Calculates distance between beads based on the number of turns in
+    the main chain. Note, here we consider distances between beads
+    not on side chains. For a particular axis, a, we calculate the
+    distance between lower_bead_ind and upper_bead_ind bead pairs,
+    delta_na = summation (k = lower_bead_ind to upper_bead_ind - 1) of (-1)^k*indica(k).
+    Args:
+        peptide: A Peptide object that includes all information about a protein.
+
+    Returns:
+        delta_n0, delta_n1, delta_n2, delta_n3: Tuple corresponding to
+                                                the number of occurrences
+                                                of turns at axes 0,1,2,3.
```

Comment:
I am not sure I fully understand what these are.. What exactly are the different `n` values? The number of turns?
Why does it go from 0 to 3? Will this always be the case? If not, maybe we should generalize this.
Since these are `defaultdict[int, ...]`, couldn't we also work with sparse matrices instead? That would also simplify the mathematical operations in `_create_distance_qubits`. Or do we not know the maximum size a priori?

Reply:
0 to 3 encodes the axis that we consider. Axes are explained in the paper, I will add this remark in the comment. Axes are fixed to be from 0 to 3 always. 
Not sure if I understand the idea of sparse matrices.
I might improve the dict data structure to resemble the final DistanceMap. In the current version I left data structures from the original code to save time since they are internal - not exposed to the user.